### ISSUE TITLE
Null-coalescing assignment operator ??= refactoring changes semantics if equality overloads exist

### ISSUE THREAD

**Issue by NoTuxNoBux** [JosefPihrt/Roslynator#852]
**Product and Version Used**: Roslynator 3.2.2.

**Steps to Reproduce**:

```csharp
class C
{
    private object _foo;
    public object Foo
    {
        get
        {
            if (_foo == null)
            {
                _foo = GetComponent<object>();
            }

            return _foo;
        }
    }

    private object _bar;
    public object Bar
    {
        get
        {
            if (_bar is null)
            {
                _bar = GetComponent<object>();
            }

            return _bar;
        }
    }
}
```

**Actual Behavior**:
```csharp
class C
{
    private object _foo;
    public object Foo
    {
        get
        {
            return _foo ??= GetComponent<object>();
        }
    }

    private object _bar;
    public object Bar
    {
        get
        {
            return _bar ??= GetComponent<object>();
        }
    }
}
```

**Expected Behavior**:
```csharp
class C
{
    private object _foo;
    public object Foo
    {
        get
        {
            if (_foo == null)
            {
                _foo = GetComponent<object>();
            }

            return _foo;
        }
    }

    private object _bar;
    public object Bar
    {
        get
        {
            return _bar ??= GetComponent<object>();
        }
    }
}
```

**Context**
This pattern using [GetComponent](https://docs.unity3d.com/ScriptReference/GameObject.GetComponent.html) occurs in Unity sometimes to make initialization lazy.

Unity overloads the `==` operator for the returned objects, making `_foo == null` behave differently from `_foo is null` (IIUC, the latter checks if the instance is directly `null`, but doesn't care about operator overloads, and the former checks if it is directly `null` _or_ deemed to be `null` by an overload). The MSDN documentation isn't entirely clear on this, but from what I can gather, `??=` uses `is null` rather than `== null`, making this refactoring potentially break code.

Plain C# example:

```csharp
using System;

public class FakeStringTest {
    string x;
    public FakeStringTest(string x){this.x=x;}
    
    public static bool operator== (FakeStringTest o1, string o2)
    {
        Console.WriteLine("Hitting overload");
        if (o2 is null) {
            return o1 is null || o1.x == null || o1.x == "";
        }
        
        return o1.x == o2;
    }
    
    public static bool operator!= (FakeStringTest o1, string o2)=> !(o1 == o2);
    public override bool Equals(object o2) {
        Console.WriteLine("HIT2");
        return true;
     }
    public override string ToString()
    {
        return $"FakeStringTest({x})";
    }
    //public override int GetHashCode() => x.GetHashCode();   
    static void Main()
    {
        FakeStringTest s1 = new FakeStringTest("Hello");
        FakeStringTest s2 = new FakeStringTest(null);
        FakeStringTest? s3 = null;
        
        Console.WriteLine("s1 is null = " + (s1 is null) + '\n');
        Console.WriteLine("s1 == null = " + (s1 == null) + '\n');
        Console.WriteLine("s1 ??= = " + (s1 ??= new FakeStringTest("Was deemed null")) + '\n');
        Console.WriteLine("s2 is null = " + (s2 is null) + '\n');
        Console.WriteLine("s2 == null = " + (s2 == null) + '\n');
        Console.WriteLine("s1 ??= = " + (s2 ??= new FakeStringTest("Was deemed null")) + '\n');
        Console.WriteLine("s3 is null = " + (s3 is null) + '\n');
        Console.WriteLine("s3 == null = " + (s3 == null) + '\n');
        Console.WriteLine("s3 ??= = " + (s3 ??= new FakeStringTest("Was deemed null")) + '\n');
    }
}
```

Output:

```
s1 is null = False

Hitting overload
s1 == null = False

s1 ??= = FakeStringTest(Hello)

s2 is null = False

Hitting overload
s2 == null = True

s1 ??= = FakeStringTest()

s3 is null = True

Hitting overload
s3 == null = True

s3 ??= = FakeStringTest(Was deemed null)
```

**Comment by JosefPihrt** [comment]
I think that `is null` is semantically same as `== null`:

https://sharplab.io/#v2:CYLg1APgAgTABAYQLACgDeq5bgBwE4CWAbgIYAuApnAPYBGAVhQMZlwD6tJeA3JtlAGYaDZqwBCXPlgwpscuAHMKZKfJnyNcAgDM4ACg5ctAZzgA7AK4AbKwEpVmuOsebDeOAF44AcWUJqALY41GYUZmQAPHSMLAB8era8si5YAL6oDppQAOzsnDyZaarpyViq+MTkVNGiPn6BwaHhEQAq8baeseYUAO7CMWQJSalAA=

**Comment by NoTuxNoBux** [comment]
Hi @JosefPihrt

Thanks for taking the time to look at this, but is it possible that this is still a problem in Roslynator 4.0.2, at least with Unity? I have a similar case to the OP:

```cs
private GameObject? foo;
private GameObject? Foo
{
    get
    {
        if (foo == null)
        {
            foo = UnityEngine.Object.FindObjectOfType<GameObject>(true);
        }

        return foo;
    }
}
```

But with VSCode extension 4.0.3 (I'm assuming this is running on Roslynator 4.0.2), I'm still seeing the suggestion to simplify:

![afbeelding](https://user-images.githubusercontent.com/73220431/157045671-fab8a373-04d7-43be-a098-bc49f20c30f5.png)

But when using code navigation to open Unity's `GameObject`, and then Unity's `Object`, these two overloads exist:

```cs
public static bool operator ==(Object x, Object y);
public static bool operator !=(Object x, Object y);
```

Meaning that the comparison with `null` above has a custom overload (FYI, Unity's code does not use nullable types yet), and this refactoring would convert that to `is null`, which has different behaviour.

Maybe there's something special about the Unity case, though?

**Comment by JosefPihrt** [comment]
Hi,

In you former code sample there was equality operator overload where parameters have different types in which case it is possible that `== null` and `is null` will behave differently.

```
public static bool operator== (FakeStringTest o1, string o2)
```

But in your latter code sample there is equality operator overload where both parameters have same type in which case does not make sense that `== null` and `is null` would behave differently

```
public static bool operator ==(Object x, Object y)
```

Source code from Unity shows that comparison with null is same whether it's `== null` or `is null`

```
public static bool operator==(Object x, Object y) { return CompareBaseObjects(x, y); }

static bool CompareBaseObjects(UnityEngine.Object lhs, UnityEngine.Object rhs)
{
    bool lhsNull = ((object)lhs) == null;
    bool rhsNull = ((object)rhs) == null;

    if (rhsNull && lhsNull) return true;

    if (rhsNull) return !IsNativeObjectAlive(lhs);
    if (lhsNull) return !IsNativeObjectAlive(rhs);

    return lhs.m_InstanceID == rhs.m_InstanceID;
}
```

https://github.com/Unity-Technologies/UnityCsReference/blob/4e6653524c3619057f7023b60c866e73f05eef40/Runtime/Export/Scripting/UnityEngineObject.bindings.cs#L103

**Comment by NoTuxNoBux** [comment]
Hi @JosefPihrt 

Thanks for taking the time to post a response.

I'm not quite following because the code seems to imply that there is in fact a difference :smile: ? I made [an example on SharpLab here using code similar to the one linked from Unity](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUH1AZgAINiAxAe0uIG89jHTMAGYgDwG4GmiLqAFKlYcAlLWAALAJYBnAHTsAvFwC+PRhtIlhANmIAjagBtilAA4BTMAENglMEqUCqNdjH40AnqLpamzACcAgBEAMKUALbmNmCWpGwW1nYOTiGi3LgBAagA7MQR0bGWAEI2spYA8gYAVpYAxsCyAu7EPpkB6llMWnx6hiZmVrb2YACEzq4cHlM+dKT5Y4UxcWUV1XWNza3txF0BvTqY+tLRxtL10sBDyaMDlMYu1MSW7HJNolr03dlBocvFBL3YyWGxQYj1crAdIdX4LYhLKIrUrlKq1BpNASvd6yDxQCDGYwZLRdXrHYEFJHFNZozaYqbGSS4zzEMBMz4/PycgIAeh5xAACmBKAYbAZjF5iHBKFB4vZiOY4gA3SxQa7SKAAMw1V3ixmo5mIYLgxAA5pZriE5MR8YSQgBaWTALwgwyWSQ2JXSSgQMAeShSawAdzkln8TD5xCuxEi0lNkmuwDAcfNYBuI1SSiNppsGuIAgA/KJ5OHGEYHsRGbIAHIE0xZgQCEV00RV3xOG112G/cumNk1uvEBtN9GNUT99tZ23GTKlqOa/P92uE4gAMlXlaZy6Jc++cOywmCIRKAckxFk0jgllkRrinbtxO5cLyxCTEEs3ey+zhc8PoRPUjnpe163vEUABveph2K+kjxLINiRHKpyWDCeBztIC4CEuda+Luc45JgR4AEpxgm9oelAJoXleUY3tOqFPr8L5jAAkjWdjSCqGwYgAgucKoCG2n6dHO6GYVW24cvue77rwhGhAAMpYmrAORxpATR1r0Y+sm8IsbHVhxXGjsAfGcZYWHssJTCkox8JVooQ5Zv2iift+micv0vbEAZRm0rx/EWVMlBSUwMlMfJISGcA5lmCZEJUsmsgyq+yamqmlhwAx+6/vklCORMxAhCEblaGVnnkqgAAsxAALK5lAAihYw4XZFMmqYEONqWEGnihAAEpYhKUNlcLtegXWyr1rihKNvy/pFHW0ZBXUhMQiD5ktWk4etxAAOQADpQHtOnPotnUdtOq27QIS2XTtG2Hcdp0RUeS0FgWWZZmtG23Z1H1Tj1fUhAA6uUUqWJYiEmtpviPUdJ3WYwf4vQe8mowcdl/iEmoTdtK7fTduPLdOcP7QjGNyW9E33QTRVEzTU4PeTz1I38OP/Z912/cTAPddNgig+DV5Q5lkHpGTT2IySeCqEAA=) which seems to confirm this.

To briefly summarize the example:
* `someUnityObject is null` does not trigger `operator==` at all, and only checks if someUnityObject is actually `null` and not an instance of any sort of object.
* `someUnityObject == null` triggers `operator==` (if someUnityObject is not actually `null`), which triggers `CompareBaseObjects`, which triggers `IsNativeObjectAlive`, which has special behaviour depending on native pointers, making its result different from `is null`. Put differently, it can return `true` even if `someUnityObject` is not `null`, but an actual instance of an object, where the native pointer inside that object is not set.
* As a result, and because `??=` uses `is null` in the background, a comparison using `==` cannot be refactored to use `??=` as it could change behaviour (comparisons using `is null` _can_ always be safely converted, and comparisons using `== null` only when there is no overload of `operator==`).

**Comment by JosefPihrt** [comment]
Ok, I seems that you are right I will provide a fix.

**Comment by JosefPihrt** [comment]
Fixed with 132dee67220a4cbf95c53d73e95dbd5d3e11d4c1
